Add sections for backend implementation and cost (#183)
SW709E19/Report: report/ChapterImplementation/Cost.tex

\end{equation}

The monthly data volume is propertional with the number of samples. 

\begin{equation}
  volume_{iot}(samples) = samples * 40\ \mathrm{KB}
\end{equation}

Configuration updates are not included in this estimate because the data volume of these are negligible in comparison to the volume of samples.

\subsection{Pub/Sub}
The cost of using Pub/Sub depends on the the volume of messages received and sent. 
The first 10 GB are free and the following costs 40 USD per TB \cite{PubSubPricing}.

\begin{equation}
  cost_{pubsub}(volume) = max(volume - 10\ \mathrm{GB}, 0) * 40\ \frac{\mathrm{USD}}{\mathrm{TB}} 
\end{equation}

The volume for the Pub/Sub is twice the data received from IoT Core because the data is also sent to the Store Sample function.

\begin{equation}
 volume_{pubsub}(samples) = 2 * volume_{iot}(samples).
\end{equation}

\subsection{Cloud Functions}
The cost of Cloud Functions depends on the number of invocations, GB-seconds, GHz-seconds, and outbound network traffic \cite{CloudFunctionsPricing}.

The first 2 million invocations are free and the following costs 0.40 USD per million invocations \cite{CloudFunctionsPricing}. 

\begin{equation}

>  cost_{invoke}(invocations) = max(invocations - 2000000, 0) * 0.40\ \frac{\mathrm{USD}}{1000000}

Reviewer: ```suggestion
 cost_{invoke}(invocations) = max(invocations - 2,000,000, 0) * 0.40\ \frac{\mathrm{USD}}{1,000,000}
```

\end{equation}

> GB-seconds is the product of time and memory used by the function, where the first 400000 GB-seconds are free and the following costs 0.0000025 USD / GB-seconds \cite{CloudFunctionsPricing}.

Reviewer: ```suggestion
GB-seconds is the product of time and memory used by the function, where the first 400,000 GB-seconds are free and the following costs 0.0000025 USD / GB-seconds \cite{CloudFunctionsPricing}.
```

\begin{equation}
  \begin{multlined}
    cost_{memory}(memory, time) = \\

>      max(memory * time - 400000\ \mathrm{GB\ s}, 0) * 0.0000025\ \frac{\mathrm{USD}}{\mathrm{GB\ s}}

Reviewer: ```suggestion
     max(memory * time - 400,000\ \mathrm{GB\ s}, 0) * 0.0000025\ \frac{\mathrm{USD}}{\mathrm{GB\ s}}
```

  \end{multlined}
\end{equation}

> GHz-seconds is the product of time and CPU used by the function, where the first 200000 GHz-seconds are free and the following costs 0.0000100 USD / GHz-seconds \cite{CloudFunctionsPricing}.

Reviewer: ```suggestion
GHz-seconds is the product of time and CPU used by the function, where the first 200,000 GHz-seconds are free and the following costs 0.0000100 USD / GHz-seconds \cite{CloudFunctionsPricing}.
```

\begin{equation}

>   cost_{cpu}(cpu, time) = max(cpu * time - 200000\ \mathrm{GHz\ s}, 0) * 0.0000100\ \frac{\mathrm{USD}}{\mathrm{GHz\ s}}

Reviewer: ```suggestion
  cost_{cpu}(cpu, time) = max(cpu * time - 200,000\ \mathrm{GHz\ s}, 0) * 0.0000100\ \frac{\mathrm{USD}}{\mathrm{GHz\ s}}
```

\end{equation}

The first 5 GB outbound network traffic is free the rest costs 0.12 USD per GB \cite{CloudFunctionsPricing}.

\begin{equation}
  cost_{network}(volume) = max(volume - 5\ \mathrm{GB}, 0) * 0.12\ \frac{\mathrm{USD}}{\mathrm{GB}}
\end{equation}

The Store Sample function runs on a 400 MHz model with 256 MB memory. 
The number of invocations equals the number of samples processed. 
Each invocation takes on average about 1 second and the outbound network volume approximates to the size of samples in JSON. 
This gives the following cost function for the Store Sample function.

\begin{equation}
  \begin{multlined}
    cost_{store}(samples) = cost_{invoke}(samples) * cost_{memory}(256\ \mathrm{MB}, samples * 1\ \mathrm{s}) \\

>      + cost_{cpu}(400\ \mathrm{MHz}, samples * 1\ \mathrm{s}) + cost_{network}(samples * 3 * 40\ \mathrm{KB})

Reviewer 2: samples that are refered to as json could be named jsonsamples instead of using * 3

  \end{multlined}
\end{equation}

The Get Samples function runs on a 1.4 GHz model with 1 GB memory. 
The number of invocations depends on the use by the frontend. 
The frontend will call the Get Samples function every second whenever it displays the graph of measurements. 

> Assuming that it is displayed 24 hours a day the number of Get Samples invocations in a month is: $60 * 60 * 24 * 30 = 2592000$.

Reviewer: ```suggestion
Assuming that it is displayed 24 hours a day the number of Get Samples invocations in a month is: $60 * 60 * 24 * 30 = 2,592,000$.
```

Each invocation takes on average about 50 milliseconds. 
The implementation of the frontend implies that measurements from a sample is only requested once. 
This gives a monthly volume of approximately the same as the Store Sample function.

\begin{equation}
  \begin{multlined}

>     cost_{get}(samples) = cost_{invoke}(2592000) * cost_{memory}(1\ \mathrm{GB}, samples * 50\ \mathrm{ms}) \\

Reviewer: ```suggestion
    cost_{get}(samples) = cost_{invoke}(2,592,000) * cost_{memory}(1\ \mathrm{GB}, samples * 50\ \mathrm{ms}) \\
```

     + cost_{cpu}(1.4\ \mathrm{GHz}, samples * 50\ \mathrm{ms}) + cost_{network}(samples * 3 * 40\ \mathrm{KB})
  \end{multlined}
\end{equation}

\subsection{Datastore}
The cost of using Datastore depends on the volume of stored data, as well as the number of reads and writes \cite{DatastorePricing}.
The first 1 GB is free and the following costs 0.18 USD per GB \cite{DatastorePricing}.

\begin{equation}
  cost_{storage}(volume) = max(volume - 1\ \mathrm{GB}, 0) * 0.18\ \frac{\mathrm{USD}}{\mathrm{GB}}
\end{equation}

> The first 50000 reads per day are free and the following cost 0.039 USD per 100000 reads \cite{DatastorePricing}.

Reviewer: ```suggestion
The first 50,000 reads per day are free and the following cost 0.039 USD per 100,000 reads \cite{DatastorePricing}.
```

\begin{equation}

>   cost_{read}(reads, days) = max(reads - 50000 * days, 0) * 0.039\ \frac{\mathrm{USD}}{100000}

Reviewer: ```suggestion
  cost_{read}(reads, days) = max(reads - 50,000 * days, 0) * 0.039\ \frac{\mathrm{USD}}{100,000}
```

\end{equation}

> The first 20000 writes per day are free and the following cost 0.117 USD per 100000 writes \cite{DatastorePricing}.

Reviewer: ```suggestion
The first 20,000 writes per day are free and the following cost 0.117 USD per 100,000 writes \cite{DatastorePricing}.
```

\begin{equation}

>   cost_{write}(writes, days) = max(writes - 20000 * days, 0) * 0.117\ \frac{\mathrm{USD}}{100000}

Reviewer: ```suggestion
  cost_{write}(writes, days) = max(writes - 20,000 * days, 0) * 0.117\ \frac{\mathrm{USD}}{100,000}
```

\end{equation}

The volume of stored data in a month is based on the number samples. 
Besides the sample data the volume also represents size of the indexes. 
To estimate this the volume of samples in JSON are multiplied by two. 
Because the implementation of the frontend implies that each sample is only read once then the number of reads equals the number of samples. 
The number of writes does also equal the number of samples.
The total cost of Datastore based on number of samples is then:

\begin{equation}

>   cost_{datastore}(samples) = cost_{storage}(2 * 3 * 40\ \mathrm{KB}) + cost_{read}(samples) + cost_{write}(samples)

Reviewer 2: *3

\end{equation}

\subsection{Total Cost}
The estimated total cost of using Google Cloud can now be derived by summing the cost functions.

\begin{equation}
  \begin{multlined}
    cost_{total}(samples) = cost_{iot}(volume_{iot}(samples)) + cost_{pubsub}(volume_{pubsub}(samples)) \\ 
    + cost_{store}(samples) + cost_{get}(samples) + cost_{datastore}(samples)
  \end{multlined}
\end{equation}

\autoref{fig:cost} shows the estimated total cost for Google Cloud given the number of samples processed by the system in one month.

\begin{figure}[H]
	\centering
	\includegraphics[width=1\textwidth]{ChapterImplementation/images/cost.png}
	\caption{The estimated cost of one month use of Google Cloud services based on number of samples.}
	\label{fig:cost}
\end{figure}

Based on \autoref{fig:cost} we can conclude that the total cost is dominated by IoT Core whereas the cost of Pub/Sub is minimal.
The cost estimation can be used to determine which components to replace in future works in order to make a cheaper solution.
The estimated cost function can be used to estimate concrete instances of the system.

> For instance a deployment of the system with 25 devices each sending a sample every minute generates $25 * 60 * 24 * 30 = 1080000$ samples with cost of approximately 267 USD for a month.

Reviewer: ```suggestion
For instance a deployment of the system with 25 devices each sending a sample every minute generates $25 * 60 * 24 * 30 = 1,080,000$ samples with cost of approximately 267 USD for a month.
```
